apps/aecore/src/aec_block_insertion.erl

```diff
@@ -228,25 +275,35 @@ build_insertion_ctx_prev(Node, PrevKeyNode) ->
                     %% Ok so the prev keyblock is present but not the prev block?
                     %% this shouldn't be the case even for the genesis block
                     {error, {illegal_orphan, node_hash(Node)}};
-                {ok, #node{type = key}} ->
-                    {error, prev_key_hash_inconsistency};
                 {ok, PrevNode} ->
-                    case node_prev_key_hash(PrevNode) =:= PrevKeyHash of
-                        true ->
+                    case {aec_chain_node:type(PrevNode), node_prev_key_hash(PrevNode) =:= PrevKeyHash} of
+                        {key, _} -> {error, prev_key_hash_inconsistency};
+                        {_, false} -> {error, prev_key_hash_inconsistency};
+                        {_, true} ->
                             %% Now assert heights
-                            build_insertion_ctx_check_prev_height(Node, PrevNode, PrevKeyNode);
-                        false ->
-                            {error, prev_key_hash_inconsistency}
+                            build_insertion_ctx_check_prev_height(Node, PrevNode, PrevKeyNode)
                     end
             end
    end.
 
-build_insertion_ctx_check_prev_height(#node{type = key} = Node, PrevNode, PrevKeyNode) ->
+-spec build_insertion_ctx_check_prev_height(
+    chain_node(), chain_node(), chain_node()) ->
+    {ok, chain_node(), chain_node()} | {error, atom()}.
+build_insertion_ctx_check_prev_height(Node, PrevNode, PrevKeyNode) ->
+    case aec_chain_node:type(Node) of
```

Comment:
Is it not simpler to work directly with the `#node{}` record? I think now we have to use `aec_chain_node` module everywhere introducing some more `case ... of`

Reply:
> Is it not simpler to work directly with the `#node{}` record? I think now we have to use `aec_chain_node` module everywhere introducing some more `case ... of`

Sure it's easier. Like Dark Side is easier than Light Side, and eating sugar for energy is easier than exercising and sleeping well :)

Reply:
Haha. I see your point, but having a top-level module does not always reduce complexity but adds more. This could introduce a tiny bit of complexity for working with the Node record. Just an opinion; your call.

Reply:
I think we should use `aec_chain_node:type` this is consistant with how things are done in other parts.

Reply:
> Haha. I see your point, but having a top-level module does not always reduce complexity but adds more. This could introduce a tiny bit of complexity for working with the Node record. Just an opinion; your call.

Actually, it reduces the overall complexity, and sometimes greatly so. Imagine someone from the core team add a field in `chain_node` or changes the order of fields. Even better, imagine something like that in a hot-swap scenario. It's scary, been there.

So basically we don't put records in .hrl and we definitely don't do cross-app matching. It's ok for a quick drafting sometimes, but never in a prod.

Reply:
@rustaem good points!

Reply:
I also agree to encapsulate protocol related stuff. Function call can be useful in mocking etc..